tests/validator_test.go

```diff
@@ -329,4 +337,611 @@ var _ = Describe("Template validator", func() {
+			vm.ObjectMeta.Annotations = map[string]string{
+				TemplateNameAnnotation:      template.Name,
+				TemplateNamespaceAnnotation: template.Namespace,
+			}
+			Expect(errors.IsInvalid(apiClient.Create(ctx, vm))).To(BeTrue(), "Should match error type because of unfulfilled validations")
+		})
+		It("test with template optional rules unfulfilled", func() {
+			template = TemplateWithRulesOptional()
+			Expect(apiClient.Create(ctx, &template)).ToNot(HaveOccurred(), "Failed to create template: %s", template.Name)
+
+			vmi = addDomainResourcesToVMI(vmi, 0, "q35", "128M")
+			vm = NewVirtualMachine(vmi)
+			vm.ObjectMeta.Annotations = map[string]string{
+				TemplateNameAnnotation:      template.Name,
+				TemplateNamespaceAnnotation: template.Namespace,
+			}
+			Expect(apiClient.Create(ctx, vm)).ToNot(HaveOccurred(), "Failed to create VM")
+		})
+		It("test with template cpu count 0 ", func() {
+			template = TemplateWithRules()
+			Expect(apiClient.Create(ctx, &template)).ToNot(HaveOccurred(), "Failed to create template: %s", template.Name)
+
+			vmi = addDomainResourcesToVMI(vmi, 0, "q35", "128M")
+			vm = NewVirtualMachine(vmi)
+			vm.ObjectMeta.Annotations = map[string]string{
+				TemplateNameAnnotation:      template.Name,
+				TemplateNamespaceAnnotation: template.Namespace,
+			}
+			Expect(apiClient.Create(ctx, vm)).ToNot(HaveOccurred(), "Failed to Create VM")
+		})
+		It("Test template with incorrect rules satisfied", func() {
+			template = TemplateWithIncorrectRules()
+			Expect(apiClient.Create(ctx, &template)).ToNot(HaveOccurred(), "Failed to create template: %s", template.Name)
+
+			vmi = addDomainResourcesToVMI(vmi, 0, "q35", "128M")
+			vm = NewVirtualMachine(vmi)
+			vm.ObjectMeta.Annotations = map[string]string{
+				TemplateNameAnnotation:      template.Name,
+				TemplateNamespaceAnnotation: template.Namespace,
+			}
+			Expect(errors.IsInvalid(apiClient.Create(ctx, vm))).To(BeTrue(), "Should have given the invalid error failing to fulfill validations")
+		})
+		It("Test template with incorrect rules unfulfilled", func() {
+			template = TemplateWithIncorrectRules()
+			Expect(apiClient.Create(ctx, &template)).ToNot(HaveOccurred(), "Failed to create template: %s", template.Name)
+
+			vmi = addDomainResourcesToVMI(vmi, 0, "q35", "32M")
+			vm = NewVirtualMachine(vmi)
+			vm.ObjectMeta.Annotations = map[string]string{
+				TemplateNameAnnotation:      template.Name,
+				TemplateNamespaceAnnotation: template.Namespace,
+			}
+			Expect(errors.IsInvalid(apiClient.Create(ctx, vm))).To(BeTrue(), "Should have given the invalid error failing to fulfill validations")
+		})
+		It("[test_id:2959] Create a VM with memory restrictions violation that succeeds with a warning", func() {
+			template = TemplateWithIncorrectRulesJustWarning()
+			Expect(apiClient.Create(ctx, &template)).ToNot(HaveOccurred(), "Failed to create template: %s", template.Name)
+
+			vmi = addDomainResourcesToVMI(vmi, 0, "q35", "1G")
```

Comment:
The VM in original test case does not have CPU info, please remove it. See my other comments why.

Reply:
Corrected